pkg/apiserver/admission/admissiontimeout/timeoutadmission.go

```diff
@@ -29,17 +31,46 @@ func (p pluginHandlerWithTimeout) Admit(ctx context.Context, a admission.Attribu
 		return nil
 	}
 
-	admissionDone := make(chan struct{})
-	admissionErr := fmt.Errorf("default to mutation error")
+	type result struct {
+		admissionErr error
+		panicErr     error
+	}
+	// if a timeout occurs, we don't want the child goroutine to hang forever
+	resultCh := make(chan result, 1)
 	go func() {
-		defer utilruntime.HandleCrash()
-		defer close(admissionDone)
-		admissionErr = mutatingHandler.Admit(ctx, a, o)
+		r := result{}
+		// NOTE: panics don't cross goroutine boundaries, so we have to handle
+		// the error here, we can't call utilruntime.HandleCrash here, then it
+		// will cause the apiserver to crash.
+		// We also need to make sure that the panic is propagated to the caller.
+		defer func() {
+			if err := recover(); err != nil {
+				// do not wrap the sentinel ErrAbortHandler panic value
+				if err != http.ErrAbortHandler {
+					// Same as stdlib http server code. Manually allocate stack
+					// trace buffer size to prevent excessively large logs
+					const size = 64 << 10
+					buf := make([]byte, size)
+					buf = buf[:runtime.Stack(buf, false)]
+					err = fmt.Sprintf("%v\n%s", err, buf)
+				}
+
+				r.panicErr = fmt.Errorf("admission panic'd: %v", err)
+				utilruntime.HandleError(r.panicErr)
+			}
+			resultCh <- r
```

Comment:
sure, but it will come at the cost of:
```
E0221 12:21:24.070372 4165137 timeoutadmission.go:61] admission panic'd: fail!
goroutine 56 [running]:
github.com/openshift/library-go/pkg/apiserver/admission/admissiontimeout.pluginHandlerWithTimeout.Admit.func1.1()
	/home/akashem/go/src/github.com/openshift/library-go/pkg/apiserver/admission/admissiontimeout/timeoutadmission.go:56 +0xbb
```

```
E0221 12:25:33.125816 4165823 timeoutadmission.go:51] admission panic'd: fail!
goroutine 14 [running]:
github.com/openshift/library-go/pkg/apiserver/admission/admissiontimeout.stack({0xa6ee40?, 0xc53e60})
	/home/akashem/go/src/github.com/openshift/library-go/pkg/apiserver/admission/admissiontimeout/timeoutadmission.go:124 +0x7d
github.com/openshift/library-go/pkg/apiserver/admission/admissiontimeout.pluginHandlerWithTimeout.Admit.func1.1()
	/home/akashem/go/src/github.com/openshift/library-go/pkg/apiserver/admission/admissiontimeout/timeoutadmission.go:50 +0x53
```


originally the first line takes you to the panic site